In [ ]:
pip install pandas

Note: you may need to restart the kernel to use updated packages.


In [ ]:
import pandas as pd
import csv



In [ ]:
#Criando um dicionário com todos os caminhos dos arquivos. Isso vai ser bom porque utilizarei as chaves para criar uma coluna de "Ano"
arquivos = {
    '2019': 'data/dados_servidores_2019.csv',
    '2020': 'data/dados_servidores_2020.csv',
    '2021': 'data/dados_servidores_2021.csv'
}

lista_dataframes = [] #Para armazenar todos os dataframes

for ano, arquivo in arquivos.items():
    try:  
        df = pd.read_csv(arquivo, encoding="utf-8")

        lista_dataframes.append(df)

    except Exception as e:
        print(f"Erro ao ler os dados do ano {ano}: {e}")


df_raw = pd.concat(lista_dataframes, ignore_index=True) #Concatenando todos os DFs



print(f"Total de linhas e colunas no df unificado: {df_raw.shape}")

C:\Users\sfrei\AppData\Local\Temp\ipykernel_6920\3536511882.py:12: DtypeWarning: Columns (26) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(arquivo, encoding="utf-8")
C:\Users\sfrei\AppData\Local\Temp\ipykernel_6920\3536511882.py:12: DtypeWarning: Columns (26) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(arquivo, encoding="utf-8")
C:\Users\sfrei\AppData\Local\Temp\ipykernel_6920\3536511882.py:12: DtypeWarning: Columns (26) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(arquivo, encoding="utf-8")


Total de linhas e colunas no df unificado: (1398332, 34)


In [ ]:
# Copia do df carregado para não precisarmos rodar a importação toda vez ue precisarmos
# voltar ao df original.
df = df_raw.copy()
df.head()

asalseanoo  asalsemess   csalseccpf  csalsematr  \
0        2019           4  ***040714**        4480   
1        2019           4  ***105614**        4529   
2        2019           4  ***507834**      512222   
3        2019           4  ***256834**      512230   
4        2019           4  ***026244**      512249   

                     nsalsenome nsalseempr   nsalsecate  nsalsecarg  \
0       EVERALDO PADRE DA SILVA       CTTU   CELETISTAS   VIGILANTE   
1             JOSE ARTUR SEABRA       CTTU   CELETISTAS   VIGILANTE   
2       VICTOR FELIX SILVA MELO       CTTU  ESTAGIARIOS  ESTAGIARIO   
3  DAIANNA M C MAIA DE OLIVEIRA       CTTU  ESTAGIARIOS  ESTAGIARIO   
4     ENILTON RUAN M DOS SANTOS       CTTU  ESTAGIARIOS  ESTAGIARIO   

       nsalsefunc  vsalsecarg  ...                        esalseunidade  \
0  SEM INFORMACAO      1198.8  ...    SETOR DE SERVICOS ADMINISTRATIVOS   
1  SEM INFORMACAO      1198.8  ...    SETOR DE SERVICOS ADMINISTRATIVOS   
2  SEM INFORMACAO         0.0  ...     SETOR DE PROCESSAMENTO DE MULTAS   
3  SEM INFORMACAO         0.0  ...  GESTAO DE UNID. DE MOB. SUSTENTAVEL   
4  SEM INFORMACAO         0.0  ...     SETOR DE PROCESSAMENTO DE MULTAS   

   dslderadmissao  dslserdesligamento  aslserjornadamensal  esalseinstrucao  \
0      1991-08-30          2021-11-01                220.0    Mdio Completo   
1      1991-09-05          2021-10-01                220.0    Mdio Completo   
2      2019-01-16          2019-07-31                110.0  Mdio Incompleto   
3      2019-02-01          2019-11-30                110.0  Mdio Incompleto   
4      2019-02-01          2020-12-31                110.0  Mdio Incompleto   

   esalsegenero  dsalseaposentadoria  esalseadministracao  eslserlotacao  \
0     Masculino                  NaN             Indireta           CTTU   
1     Masculino                  NaN             Indireta           CTTU   
2     Masculino                  NaN             Indireta           CTTU   
3      Feminino                  NaN             Indireta           CTTU   
4     Masculino                  NaN             Indireta           CTTU   

        eselsesituacao  
0  Desligado/Exonerado  
1  Desligado/Exonerado  
2  Desligado/Exonerado  
3  Desligado/Exonerado  
4  Desligado/Exonerado  

[5 rows x 34 columns]

In [ ]:
dict_dados = [
      {
        "codigo": "asalseanoo",
        "descricao": "ANO",
      },
      {
        "codigo": "asalsemess",
        "descricao": "MES",

      },
      {
        "codigo": "csalseccpf",
        "descricao": "CPF",

      },
      {
        "codigo": "csalsematr",
        "descricao": "MATRICULA",

      },
      {
        "codigo": "nsalsenome",
        "descricao": "NOME",

      },
      {
        "codigo": "nsalseempr",
        "descricao": "SECRETARIA/ORGAO",

      },
      {
        "codigo": "nsalsecate",
        "descricao": "CATEGORIA",

      },
      {
        "codigo": "nsalsecarg",
        "descricao": "CARGO",

      },
      {
        "codigo": "nsalsefunc",
        "descricao": "FUNCAO",

      },
      {
        "codigo": "vsalsecarg",
        "descricao": "VENCIMENTO CARGO",

      },
      {
        "codigo": "vsalsefunc",
        "descricao": "VENCIMENTO FUNCAO",

      },
      {
        "codigo": "vsalseremu",
        "descricao": "REMUNERACAO",

      },
      {
        "codigo": "vsalseferi",
        "descricao": "VENCIMENTO FERIAS",

      },
      {
        "codigo": "vsalsenatl",
        "descricao": "VENCIMENTO NATAL",

      },
      {
        "codigo": "vsalseoutr",
        "descricao": "VENCIMENTOS OUTROS",

      },
      {
        "codigo": "vsalseprov",
        "descricao": "TOTAL PROVENTOS",

      },
      {
        "codigo": "vsalsedxcd",
        "descricao": "DESCONTOS EXCEDENTES",

      },
      {
        "codigo": "vsalsedrst",
        "descricao": "DESCONTO FALTA OU RESTITUICAO",

      },
      {
        "codigo": "vsalsedprv",
        "descricao": "DESCONTO PREVIDENCIA",

      },
      {
        "codigo": "vsalsedrrf",
        "descricao": "DESCONTO IRRF",

      },
      {
        "codigo": "vsalsedtot",
        "descricao": "DESCONTOS COMPULSORIOS",

      },
      {
        "codigo": "vsalseliqd",
        "descricao": "VENCIMENTO LIQUIDO",

      },
      {
        "codigo": "vsalsedife",
        "descricao": "DIFERENCA VENCIMENTO MES",

      },
      {
        "codigo": "tslserulat",
        "descricao": "DATA E HORA EVENTO",

      },
      {
        "codigo": "eslserlotacao",
        "descricao": "LOTACAO",

      },
      {
        "codigo": "dslderadmissao",
        "descricao": "DATA ADMISSAO",

      },
      {
        "codigo": "dslserdesligamento",
        "descricao": "DATA DESLIGAMENTO",

      },
      {
        "codigo": "aslserjornadamensal",
        "descricao": "JORANADA MENSAL",

      },
      {
          "codigo" : "esalseinstrucao" ,
          "descricao": "GRAU DE INSTRUCAO"
      },
      {
          "codigo" : "esalsegenero" ,
          "descricao": "GENERO"
      },
      {
          "codigo" : "eselsesituacao" ,
          "descricao": "SITUACAO"
      },
      {
          "codigo" : "esalseadministracao" ,
          "descricao": "TIPO ADM"
      }

      ]

### Padronizando as Colunas

#### Identificando Colunas

In [ ]:
# Renomeando as colunas com base nas discrições do dicionario de dados para nomes mais intuitivos
for campo in dict_dados:
    df = df.rename(columns={campo['codigo']: campo['descricao']})

In [ ]:
print('DataSet')
display(df.head(5))
print('Resumo de informações:')
display(df.info())
print('Verificando Valores únicos:')
display(df.nunique(dropna=False))
print('Verificando valores nulos:')
display(df.isnull().sum())

DataSet


ANO  MES          CPF  MATRICULA                          NOME  \
0  2019    4  ***040714**       4480       EVERALDO PADRE DA SILVA   
1  2019    4  ***105614**       4529             JOSE ARTUR SEABRA   
2  2019    4  ***507834**     512222       VICTOR FELIX SILVA MELO   
3  2019    4  ***256834**     512230  DAIANNA M C MAIA DE OLIVEIRA   
4  2019    4  ***026244**     512249     ENILTON RUAN M DOS SANTOS   

  SECRETARIA/ORGAO    CATEGORIA       CARGO          FUNCAO  VENCIMENTO CARGO  \
0             CTTU   CELETISTAS   VIGILANTE  SEM INFORMACAO            1198.8   
1             CTTU   CELETISTAS   VIGILANTE  SEM INFORMACAO            1198.8   
2             CTTU  ESTAGIARIOS  ESTAGIARIO  SEM INFORMACAO               0.0   
3             CTTU  ESTAGIARIOS  ESTAGIARIO  SEM INFORMACAO               0.0   
4             CTTU  ESTAGIARIOS  ESTAGIARIO  SEM INFORMACAO               0.0   

   ...                        esalseunidade  DATA ADMISSAO  DATA DESLIGAMENTO  \
0  ...    SETOR DE SERVICOS ADMINISTRATIVOS     1991-08-30         2021-11-01   
1  ...    SETOR DE SERVICOS ADMINISTRATIVOS     1991-09-05         2021-10-01   
2  ...     SETOR DE PROCESSAMENTO DE MULTAS     2019-01-16         2019-07-31   
3  ...  GESTAO DE UNID. DE MOB. SUSTENTAVEL     2019-02-01         2019-11-30   
4  ...     SETOR DE PROCESSAMENTO DE MULTAS     2019-02-01         2020-12-31   

   JORANADA MENSAL  GRAU DE INSTRUCAO     GENERO  dsalseaposentadoria  \
0            220.0      Mdio Completo  Masculino                  NaN   
1            220.0      Mdio Completo  Masculino                  NaN   
2            110.0    Mdio Incompleto  Masculino                  NaN   
3            110.0    Mdio Incompleto   Feminino                  NaN   
4            110.0    Mdio Incompleto  Masculino                  NaN   

   TIPO ADM  LOTACAO             SITUACAO  
0  Indireta     CTTU  Desligado/Exonerado  
1  Indireta     CTTU  Desligado/Exonerado  
2  Indireta     CTTU  Desligado/Exonerado  
3  Indireta     CTTU  Desligado/Exonerado  
4  Indireta     CTTU  Desligado/Exonerado  

[5 rows x 34 columns]

Resumo de informações:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1398332 entries, 0 to 1398331
Data columns (total 34 columns):
 #   Column                         Non-Null Count    Dtype  
---  ------                         --------------    -----  
 0   ANO                            1398332 non-null  int64  
 1   MES                            1398332 non-null  int64  
 2   CPF                            1398332 non-null  object 
 3   MATRICULA                      1398332 non-null  int64  
 4   NOME                           1398332 non-null  object 
 5   SECRETARIA/ORGAO               1398332 non-null  object 
 6   CATEGORIA                      1398332 non-null  object 
 7   CARGO                          1398332 non-null  object 
 8   FUNCAO                         1398332 non-null  object 
 9   VENCIMENTO CARGO               1398332 non-null  float64
 10  VENCIMENTO FUNCAO              1398332 non-null  float64
 11  REMUNERACAO                    1398332 non-null  floa

None

Verificando Valores únicos:


ANO                                   3
MES                                  12
CPF                               35744
MATRICULA                         47933
NOME                              49371
SECRETARIA/ORGAO                     10
CATEGORIA                            15
CARGO                               613
FUNCAO                              126
VENCIMENTO CARGO                   5726
VENCIMENTO FUNCAO                   331
REMUNERACAO                       25462
VENCIMENTO FERIAS                 14472
VENCIMENTO NATAL                  36912
VENCIMENTOS OUTROS                63159
TOTAL PROVENTOS                  134297
DESCONTOS EXCEDENTES                704
DESCONTO FALTA OU RESTITUICAO      5507
DESCONTO PREVIDENCIA              22247
DESCONTO IRRF                     39108
DESCONTOS COMPULSORIOS            52618
VENCIMENTO LIQUIDO               121022
DIFERENCA VENCIMENTO MES          20767
DATA E HORA EVENTO                    3
esalseunidade                      1564


Verificando valores nulos:


ANO                                    0
MES                                    0
CPF                                    0
MATRICULA                              0
NOME                                   0
SECRETARIA/ORGAO                       0
CATEGORIA                              0
CARGO                                  0
FUNCAO                                 0
VENCIMENTO CARGO                       0
VENCIMENTO FUNCAO                      0
REMUNERACAO                            0
VENCIMENTO FERIAS                      0
VENCIMENTO NATAL                       0
VENCIMENTOS OUTROS                     0
TOTAL PROVENTOS                        0
DESCONTOS EXCEDENTES                   0
DESCONTO FALTA OU RESTITUICAO          0
DESCONTO PREVIDENCIA                   0
DESCONTO IRRF                          0
DESCONTOS COMPULSORIOS                 0
VENCIMENTO LIQUIDO                     0
DIFERENCA VENCIMENTO MES               0
DATA E HORA EVENTO                     0
esalseunidade   

### Algumas colunas não foram identificadas no dicionario de dados disponivel, vamos tentar identifica-las e analisar quais fazem sentido para nossa analise e quais podemos descartar.

In [ ]:
print('Coluna esalseunidade:')
display(df['esalseunidade'].unique())
print('Coluna dsalseaposentadoria:') 
display(df['dsalseaposentadoria'].unique())


Coluna esalseunidade:


array(['SETOR DE SERVICOS ADMINISTRATIVOS',
       'SETOR DE PROCESSAMENTO DE MULTAS',
       'GESTAO DE UNID. DE MOB. SUSTENTAVEL', ...,
       'SEC. DE MEIO AMBIENTE E SUSTENTABILIDADE',
       'UNIDADE DE ANALISE DAS LICIT DE OBRAS E',
       'FUND. DE CULTURA CIDADE DO RECIFE'], shape=(1564,), dtype=object)

Coluna dsalseaposentadoria:


array([nan, '2013-10-18', '2013-12-22', ..., '2021-10-02', '2021-08-05',
       '2021-06-29'], shape=(2741,), dtype=object)

### Vamos Analisar A coluna 'esalseunidade', verificamos uma possivel correlação com outras colunas e queremos entender se podemos otimizar a visão do nosso dataset

In [ ]:
print(df['esalseunidade'].unique())
print("")
print(df['SECRETARIA/ORGAO'].unique())
print("")
print(df['LOTACAO'].unique())

['SETOR DE SERVICOS ADMINISTRATIVOS' 'SETOR DE PROCESSAMENTO DE MULTAS'
 'GESTAO DE UNID. DE MOB. SUSTENTAVEL' ...
 'SEC. DE MEIO AMBIENTE E SUSTENTABILIDADE'
 'UNIDADE DE ANALISE DAS LICIT DE OBRAS E'
 'FUND. DE CULTURA CIDADE DO RECIFE']

['CTTU' 'URB RECIFE' 'FUND. DE CULTURA CIDADE RECIFE'
 'CIA. SERV. URBANO DO RECIFE' 'EMPRESA MUNIC. DE INFORMATICA'
 'PREFEITURA DA CIDADE DO RECIFE' 'FUNDO RECIFIN'
 'EMPRESA MANUT. LIMPEZA URBANA' 'FUNDO RECIPREV'
 'RECIPREV-AUT.PREV.MUN.RECIFE']

['CTTU' 'URB' 'SEC DE PLANEJAMENTO URBANO' 'GABINETE DE IMPRENSA'
 'FUN. CULTURA' 'CSURB' 'SEC DE POLITICA URBANA E LICENCIAMENTO' 'EMPREL'
 'CMR' 'EMLURB' 'SEC DE EDUCACAO' 'SEC DE SAUDE' 'RECIPREV'
 'SEC DE MEIO AMBIENTE E SUSTENTABILIDADE' 'SEC DE SEGURANCA CIDADA'
 'GABINETE DO PREFEITO' 'FUN. RECIFIN' 'FUN. RECIPREV'
 'SEC PLANEJAMENTO GESTAO TRANSFOR DIGITAL' 'SEC DE TURISMO E LAZER'
 'SEC DE SANEAMENTO' 'GABINETE DA VICE PREFEITA' 'SEC DA MULHER'
 'SEC DESENV SOCIAL DIR HUMAN JUVENTU DROG'
 'ASSE


é possivel perceber que 'SECRETARIA/ORGAO' está contido em 'LOTACAO' e que 'esalseunidade' é tão especifico com 1564 valores unicos que fica dificil visualizar a utilidade prática em nossas análises.

Conclusão: Iremos dropar as colunas SECRETARIA/ORGAO e esalseunide.

In [ ]:
#df = df.drop(columns=['SECRETARIA/ORGAO', 'esalseunidade'], axis=1) #Removendo as colunas
# Comentamos essa linha para manter as colunas necessárias para a modelagem dimensional

ANO  MES          CPF  MATRICULA                          NOME  \
0  2019    4  ***040714**       4480       EVERALDO PADRE DA SILVA   
1  2019    4  ***105614**       4529             JOSE ARTUR SEABRA   
2  2019    4  ***507834**     512222       VICTOR FELIX SILVA MELO   
3  2019    4  ***256834**     512230  DAIANNA M C MAIA DE OLIVEIRA   
4  2019    4  ***026244**     512249     ENILTON RUAN M DOS SANTOS   

     CATEGORIA       CARGO          FUNCAO  VENCIMENTO CARGO  \
0   CELETISTAS   VIGILANTE  SEM INFORMACAO            1198.8   
1   CELETISTAS   VIGILANTE  SEM INFORMACAO            1198.8   
2  ESTAGIARIOS  ESTAGIARIO  SEM INFORMACAO               0.0   
3  ESTAGIARIOS  ESTAGIARIO  SEM INFORMACAO               0.0   
4  ESTAGIARIOS  ESTAGIARIO  SEM INFORMACAO               0.0   

   VENCIMENTO FUNCAO  ...   DATA E HORA EVENTO  DATA ADMISSAO  \
0                0.0  ...  2024-01-11 17:27:56     1991-08-30   
1                0.0  ...  2024-01-11 17:27:56     1991-09-05   
2                0.0  ...  2024-01-11 17:27:56     2019-01-16   
3                0.0  ...  2024-01-11 17:27:56     2019-02-01   
4                0.0  ...  2024-01-11 17:27:56     2019-02-01   

   DATA DESLIGAMENTO  JORANADA MENSAL  GRAU DE INSTRUCAO     GENERO  \
0         2021-11-01            220.0      Mdio Completo  Masculino   
1         2021-10-01            220.0      Mdio Completo  Masculino   
2         2019-07-31            110.0    Mdio Incompleto  Masculino   
3         2019-11-30            110.0    Mdio Incompleto   Feminino   
4         2020-12-31            110.0    Mdio Incompleto  Masculino   

   dsalseaposentadoria  TIPO ADM  LOTACAO             SITUACAO  
0                  NaN  Indireta     CTTU  Desligado/Exonerado  
1                  NaN  Indireta     CTTU  Desligado/Exonerado  
2                  NaN  Indireta     CTTU  Desligado/Exonerado  
3                  NaN  Indireta     CTTU  Desligado/Exonerado  
4                  NaN  Indireta     CTTU  Desligado/Exonerado  

[5 rows x 32 columns]

sobre a coluna dsalseaposentadoria: Não conseguimos indentificar com clareza ao que se refere a data e decidimos dropar a coluna

In [ ]:
df = df.drop(columns='dsalseaposentadoria')


## Tratando valores nulos

In [ ]:
df.isnull().sum()

ANO                                    0
MES                                    0
CPF                                    0
MATRICULA                              0
NOME                                   0
CATEGORIA                              0
CARGO                                  0
FUNCAO                                 0
VENCIMENTO CARGO                       0
VENCIMENTO FUNCAO                      0
REMUNERACAO                            0
VENCIMENTO FERIAS                      0
VENCIMENTO NATAL                       0
VENCIMENTOS OUTROS                     0
TOTAL PROVENTOS                        0
DESCONTOS EXCEDENTES                   0
DESCONTO FALTA OU RESTITUICAO          0
DESCONTO PREVIDENCIA                   0
DESCONTO IRRF                          0
DESCONTOS COMPULSORIOS                 0
VENCIMENTO LIQUIDO                     0
DIFERENCA VENCIMENTO MES               0
DATA E HORA EVENTO                     0
DATA ADMISSAO                         36
DATA DESLIGAMENT

In [ ]:
checkpoint_1 = df.copy()

verificando os valores que foram identificados como null 

- DATA ADMISSAO e JORANADA MENSAL: temos 36 linhas NaN
- GENERO: Apenas 2 linhas
- GRAU DE INSTRUCAO

In [ ]:
print("Valores únicos da coluna 'JORANADA MENSAL': \n", df['JORANADA MENSAL'].unique())
print("\nValores únicos da coluna 'DATA ADMISSAO': \n", df['DATA ADMISSAO'].unique())
print("\nValores unicos da coluna 'GRAU DE INSTRUCAO': \n", df['GRAU DE INSTRUCAO'].unique())
print("\nValores unicos da coluna 'GENERO': \n", df['GENERO'].unique())

Valores únicos da coluna 'JORANADA MENSAL': 
 [220. 110. 120. 200. 180. 145.   0. 270. 150. 160.  80. 100.  nan 240.
 250. 125. 260. 190. 140. 210. 115. 235. 255. 265.  32. 165. 195. 170.
 130. 225.  18.  20. 108.  40.   8. 185. 155. 230. 245. 135. 175. 300.
 147. 144. 215. 205.  10.]

Valores únicos da coluna 'DATA ADMISSAO': 
 ['1991-08-30' '1991-09-05' '2019-01-16' ... '2021-12-10' '2021-12-20'
 '2012-12-12']

Valores unicos da coluna 'GRAU DE INSTRUCAO': 
 ['Mdio Completo' 'Mdio Incompleto' 'Superior Incompleto'
 'Fundamental 2 Incompleto' 'Ps-graduao Completo' 'Superior Completo'
 'Fundamental 2 Completo' 'Mestrado Completo' 'Analfabeto'
 'Fundamental 1 Completo' 'Fundamental 1 Incompleto' 'Indefinido' nan
 'Doutorado Completo']

Valores unicos da coluna 'GENERO': 
 ['Masculino' 'Feminino' nan 'NI']


Com base no critério de impacto, e por padronização do domínio das variaveis, decidimos retirar as linhas nulas de DATA ADMISSAO e JORANADA MENSAL.

In [ ]:
df = df.dropna(subset=['DATA ADMISSAO'])
df = df.dropna(subset=['JORANADA MENSAL'])

No caso de GENERO e GRAU DE INSTRUCAO temos definido uma categoria para valores indefinidos ou não informados. vamos fazer essas categorias absorverem os NaN sem problemas de interferir em nossa análise.

Sobre a coluna DATA DESLIGAMENTO, entendemos que os nulos se referem as servidores ativos que realmente não possuem data de desligamento, então criaremos de forma especial uma categoria chamada "NP" (Não se aplica).

Agora, vamos tratar preencher os valores nulos com os valores que definimos.

In [ ]:
df['GENERO'] = df['GENERO'].fillna('NI')
df['GRAU DE INSTRUCAO'] = df['GRAU DE INSTRUCAO'].fillna('Indefinido')
df['DATA DESLIGAMENTO'] = df['DATA DESLIGAMENTO'].fillna('NP')

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1398296 entries, 0 to 1398331
Data columns (total 31 columns):
 #   Column                         Non-Null Count    Dtype  
---  ------                         --------------    -----  
 0   ANO                            1398296 non-null  int64  
 1   MES                            1398296 non-null  int64  
 2   CPF                            1398296 non-null  object 
 3   MATRICULA                      1398296 non-null  int64  
 4   NOME                           1398296 non-null  object 
 5   CATEGORIA                      1398296 non-null  object 
 6   CARGO                          1398296 non-null  object 
 7   FUNCAO                         1398296 non-null  object 
 8   VENCIMENTO CARGO               1398296 non-null  float64
 9   VENCIMENTO FUNCAO              1398296 non-null  float64
 10  REMUNERACAO                    1398296 non-null  float64
 11  VENCIMENTO FERIAS              1398296 non-null  float64
 12  VENCIMENTO NATAL   

Análise Final sobre Colunas:

Vamos remover colunas que passam informaçãoes muita especificas, mas desnecessarias para uma análise estatística quantitativa e generalizada que pretendemos fazer. 

Exemplos: Colunas de valor salarial que passam grande parte do dataset zeradas.


In [ ]:
# Vencimentos, Remuneração, Proventos
print('quantidades de linhas em que REMUNERACAO é zero:', sum(df['REMUNERACAO'] == 0))
print('\nquantidades de linhas em que VENCIMENTO FUNCAO é zero:', sum(df['VENCIMENTO FUNCAO'] == 0))
print('\nquantidades de linhas em que VENCIMENTO CARGO é zero:', sum(df['VENCIMENTO CARGO'] == 0))
print('\nquantidades de linhas em que VENCIMENTO FERIAS é zero:', sum(df['VENCIMENTO FERIAS'] == 0))
print('\nquantidades de linhas em que VENCIMENTO NATAL é zero:', sum(df['VENCIMENTO NATAL'] == 0))
print('\nquantidades de linhas em que VENCIMENTOS OUTROS é zero:', sum(df['VENCIMENTOS OUTROS'] == 0))
print('\nquantidades de linhas em que DIFERENCA VENCIMENTO MES é zero:', sum(df['DIFERENCA VENCIMENTO MES'] == 0))
print('\nquantidades de linhas em que VENCIMENTO LIQUIDO é zero:', sum(df['VENCIMENTO LIQUIDO'] == 0))

# Descontos
print('\nquantidades de linhas em que DESCONTOS EXCEDENTES é zero:', sum(df['DESCONTOS EXCEDENTES'] == 0))
print('\nquantidades de linhas em que DESCONTO FALTA OU RESTITUICAO é zero:', sum(df['DESCONTO FALTA OU RESTITUICAO'] == 0))
print('\nquantidades de linhas em que DESCONTO PREVIDENCIA é zero:', sum(df['DESCONTO PREVIDENCIA'] == 0))
print('\nquantidades de linhas em que DESCONTO IRRF é zero:', sum(df['DESCONTO IRRF'] == 0))
print('\nquantidades de linhas em que DESCONTOS COMPULSORIOS é zero:', sum(df['DESCONTOS COMPULSORIOS'] == 0))

quantidades de linhas em que REMUNERACAO é zero: 292118

quantidades de linhas em que VENCIMENTO FUNCAO é zero: 1326267

quantidades de linhas em que VENCIMENTO CARGO é zero: 354832

quantidades de linhas em que VENCIMENTO FERIAS é zero: 1326372

quantidades de linhas em que VENCIMENTO NATAL é zero: 1299059

quantidades de linhas em que VENCIMENTOS OUTROS é zero: 121471

quantidades de linhas em que DIFERENCA VENCIMENTO MES é zero: 1224113

quantidades de linhas em que VENCIMENTO LIQUIDO é zero: 6746

quantidades de linhas em que DESCONTOS EXCEDENTES é zero: 1392451

quantidades de linhas em que DESCONTO FALTA OU RESTITUICAO é zero: 1370779

quantidades de linhas em que DESCONTO PREVIDENCIA é zero: 338260

quantidades de linhas em que DESCONTO IRRF é zero: 571077

quantidades de linhas em que DESCONTOS COMPULSORIOS é zero: 267561


percebemos que algumas colunas passam em torno de 90% do total de linhas, zeradas. Optamos por remover essas colunas para deixar nosso dataset mais limpo e objetivo.



In [ ]:
checkpoint_2 = df.copy()

In [ ]:
col_drop = ['VENCIMENTO FUNCAO', 'VENCIMENTO FERIAS', 'VENCIMENTO NATAL', 'DIFERENCA VENCIMENTO MES', 'DESCONTOS EXCEDENTES', 'DESCONTO FALTA OU RESTITUICAO']

df = df.drop(columns=col_drop, axis=1) #Removendo as colunas

RESULTADO FINAL

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1398296 entries, 0 to 1398331
Data columns (total 25 columns):
 #   Column                  Non-Null Count    Dtype  
---  ------                  --------------    -----  
 0   ANO                     1398296 non-null  int64  
 1   MES                     1398296 non-null  int64  
 2   CPF                     1398296 non-null  object 
 3   MATRICULA               1398296 non-null  int64  
 4   NOME                    1398296 non-null  object 
 5   CATEGORIA               1398296 non-null  object 
 6   CARGO                   1398296 non-null  object 
 7   FUNCAO                  1398296 non-null  object 
 8   VENCIMENTO CARGO        1398296 non-null  float64
 9   REMUNERACAO             1398296 non-null  float64
 10  VENCIMENTOS OUTROS      1398296 non-null  float64
 11  TOTAL PROVENTOS         1398296 non-null  float64
 12  DESCONTO PREVIDENCIA    1398296 non-null  float64
 13  DESCONTO IRRF           1398296 non-null  float64
 14  DESCONT

## Etapa de Modelagem e Carga (Load)

### Configuração da Conexão com o Banco de Dados
Agora que os dados estão limpos, vamos configurar a conexão com o PostgreSQL utilizando as variáveis de ambiente definidas no arquivo `.env`. Isso nos permitirá carregar as tabelas finais.



In [ ]:
import os
from dotenv import load_dotenv
from sqlalchemy import create_engine

# Carrega variáveis do arquivo .env
load_dotenv()

# Configura conexão
db_user = os.getenv('POSTGRES_USER')
db_password = os.getenv('POSTGRES_PASSWORD')
db_host = os.getenv('POSTGRES_HOST')
db_port = os.getenv('POSTGRES_PORT')
db_name = os.getenv('POSTGRES_DB')

DATABASE_URL = f"postgresql://{db_user}:{db_password}@{db_host}:{db_port}/{db_name}"
engine = create_engine(DATABASE_URL)

print("Engine de banco de dados criada.")

### Criando Dimensões Simples
Vamos criar as primeiras tabelas de dimensão do nosso modelo estrela.
* **dim_tempo:** Baseada nos anos e meses únicos.
* **dim_cargo:** Baseada na combinação única de Cargo, Função e Categoria.
* **dim_situacao:** Mapeia se o servidor está ativo ou desligado.

In [ ]:
# --- Dimensão Tempo ---
dim_tempo = df[['ANO', 'MES']].drop_duplicates().sort_values(['ANO', 'MES']).reset_index(drop=True)
dim_tempo['id_tempo'] = dim_tempo['ANO'] * 100 + dim_tempo['MES']
dim_tempo['semestre'] = dim_tempo['MES'].apply(lambda x: 1 if x <= 6 else 2)
dim_tempo['trimestre'] = dim_tempo['MES'].apply(lambda x: (x - 1) // 3 + 1)
meses = {1:'JANEIRO', 2:'FEVEREIRO', 3:'MARCO', 4:'ABRIL', 5:'MAIO', 6:'JUNHO', 
         7:'JULHO', 8:'AGOSTO', 9:'SETEMBRO', 10:'OUTUBRO', 11:'NOVEMBRO', 12:'DEZEMBRO'}
dim_tempo['nome_mes'] = dim_tempo['MES'].map(meses)

# --- Dimensão Cargo ---
cols_cargo = ['CARGO', 'FUNCAO', 'CATEGORIA']
dim_cargo = df[cols_cargo].drop_duplicates().sort_values(cols_cargo).reset_index(drop=True)
dim_cargo['id_cargo'] = dim_cargo.index + 1

# --- Dimensão Situação ---
dim_situacao = df[['SITUACAO']].drop_duplicates().sort_values('SITUACAO').reset_index(drop=True)
dim_situacao['id_situacao'] = dim_situacao.index + 1
dim_situacao['is_ativo'] = ~dim_situacao['SITUACAO'].str.contains('DESLIGADO/EXONERADO|APOSENTADO', na=False, case=False)

print("Dimensões Tempo, Cargo e Situação preparadas.")

### Criando Dimensão Lotação
Aqui normalizamos os dados de local de trabalho. Renomeamos as colunas originais (`esalseunidade`, `SECRETARIA/ORGAO`) para o padrão do banco (`unidade_nome`, `entidade_nome`) e criamos um ID único para cada lotação.


In [ ]:
# --- Dimensão Lotação ---
cols_lotacao = ['LOTACAO', 'esalseunidade', 'SECRETARIA/ORGAO', 'TIPO ADM']
col_rename_lotacao = {
    'esalseunidade': 'unidade_nome', 
    'SECRETARIA/ORGAO': 'entidade_nome',
    'LOTACAO': 'lotacao_nome',
    'TIPO ADM': 'tipo_administracao'
}

dim_lotacao = df[cols_lotacao].drop_duplicates()
dim_lotacao = dim_lotacao.rename(columns=col_rename_lotacao)
dim_lotacao = dim_lotacao.sort_values(['entidade_nome', 'unidade_nome', 'lotacao_nome']).reset_index(drop=True)
dim_lotacao['id_lotacao'] = dim_lotacao.index + 1

print(f"Dimensão Lotação criada: {len(dim_lotacao)} registros.")

### Criando Dimensão Servidor
Esta etapa aplica a lógica de deduplicação. Agrupamos os registros por Matrícula, Entidade e CPF para garantir que cada servidor tenha um registro único na dimensão, preservando a data de admissão mais antiga e os dados cadastrais mais recentes.

In [ ]:
# --- Dimensão Servidor ---
cols_group = ['MATRICULA', 'SECRETARIA/ORGAO', 'CPF']

dim_servidor = df.groupby(cols_group).agg({
    'NOME': 'max',
    'GENERO': 'max',
    'GRAU DE INSTRUCAO': 'max',
    'DATA ADMISSAO': 'min',
    'DATA DESLIGAMENTO': 'max'
}).reset_index()

dim_servidor = dim_servidor.rename(columns={'SECRETARIA/ORGAO': 'entidade_nome'})
dim_servidor['id_servidor'] = dim_servidor.index + 1

print(f"Dimensão Servidor criada: {len(dim_servidor)} registros.")

### Construindo a Tabela Fato (fato_folha)
Realizamos o cruzamento (merge) do dataset principal com as dimensões criadas anteriormente. O objetivo é substituir os textos descritivos pelos IDs numéricos (Foreign Keys) e manter apenas as colunas de métricas financeiras e horas trabalhadas.

In [ ]:
# --- Fato Folha ---
fato_folha = df.copy()

# Joins para obter os IDs
fato_folha['id_tempo'] = fato_folha['ANO'] * 100 + fato_folha['MES']
fato_folha = pd.merge(fato_folha, dim_cargo, on=['CARGO', 'FUNCAO', 'CATEGORIA'], how='left')
fato_folha = pd.merge(fato_folha, dim_situacao, on='SITUACAO', how='left')

# Join Lotação
fato_folha = pd.merge(fato_folha, 
                      dim_lotacao.rename(columns={'unidade_nome':'esalseunidade', 'entidade_nome':'SECRETARIA/ORGAO', 'lotacao_nome':'LOTACAO', 'tipo_administracao':'TIPO ADM'}),
                      on=['LOTACAO', 'esalseunidade', 'SECRETARIA/ORGAO', 'TIPO ADM'], 
                      how='left')

# Join Servidor
fato_folha = pd.merge(fato_folha, 
                      dim_servidor.rename(columns={'entidade_nome':'SECRETARIA/ORGAO'}),
                      on=['MATRICULA', 'SECRETARIA/ORGAO', 'CPF'], 
                      how='left')

# Seleção e renomeação final
colunas_finais = [
    'id_servidor', 'id_cargo', 'id_lotacao', 'id_tempo', 'id_situacao',
    'JORANADA MENSAL', 'VENCIMENTO CARGO', 'REMUNERACAO', 'VENCIMENTO LIQUIDO',
    'DESCONTO PREVIDENCIA', 'DESCONTO IRRF', 'TOTAL PROVENTOS'
]

mapa_colunas_fato = {
    'JORANADA MENSAL': 'jornada_mensal',
    'VENCIMENTO CARGO': 'valor_salario_base',
    'REMUNERACAO': 'valor_remuneracao_bruta',
    'VENCIMENTO LIQUIDO': 'valor_liquido',
    'DESCONTO PREVIDENCIA': 'valor_previdencia',
    'DESCONTO IRRF': 'valor_irrf',
    'TOTAL PROVENTOS': 'valor_total_proventos'
}

fato_folha_final = fato_folha[colunas_finais].rename(columns=mapa_colunas_fato)
print("Fato Folha montada.")

### Carga Final no Data Warehouse
Por fim, enviamos as tabelas modeladas (`dim_` e `fato_`) para o banco de dados PostgreSQL. Utilizamos a engine do SQLAlchemy para realizar a inserção dos dados.

In [ ]:
def carregar_tabela(df, nome_tabela, engine, schema='public'):
    print(f"Carregando tabela {nome_tabela}...")
    df.to_sql(nome_tabela, con=engine, if_exists='replace', index=False, schema=schema)
    print(f"Tabela {nome_tabela} carregada com sucesso!")

# Se quiser usar um schema diferente, altere 'public' para o nome do seu schema (ex: 'dw_python')
# Mas lembre-se de criar o schema no banco antes se ele não existir
schema_destino = 'public' 

carregar_tabela(dim_tempo, 'dim_tempo', engine, schema_destino)
carregar_tabela(dim_cargo, 'dim_cargo', engine, schema_destino)
carregar_tabela(dim_situacao, 'dim_situacao', engine, schema_destino)
carregar_tabela(dim_lotacao, 'dim_lotacao', engine, schema_destino)
carregar_tabela(dim_servidor, 'dim_servidor', engine, schema_destino)
carregar_tabela(fato_folha_final, 'fato_folha', engine, schema_destino)

print("ETL concluído com sucesso!")

# 📊 Resultados do Pipeline ETL

Com a execução deste notebook, concluímos com sucesso a transformação dos dados brutos de servidores (2019-2021) em um **Data Warehouse** modelado em Esquema Estrela.

## 1. Arquitetura Gerada
A abordagem *Python-Driven* replicou a lógica de modelagem dimensional, resultando na seguinte estrutura carregada no PostgreSQL:

* **Fato:**
    * `fato_folha`: Tabela central contendo métricas financeiras (salários, descontos) e chaves estrangeiras. Granularidade: Um registro por servidor por mês.
* **Dimensões:**
    * `dim_servidor`: Cadastro único de servidores (Deduplicado por Matrícula + CPF + Entidade).
    * `dim_lotacao`: Normalização da hierarquia organizacional (Entidade > Unidade > Lotação).
    * `dim_cargo`: Combinações únicas de Cargo, Função e Categoria.
    * `dim_situacao`: Status funcional (Ativo, Aposentado, Exonerado) com flag booleana.
    * `dim_tempo`: Calendário auxiliar derivado das folhas de pagamento.

## 2. Resumo das Transformações
As principais etapas de tratamento aplicadas foram:
1.  **Limpeza (Cleaning):** Remoção de colunas zeradas e linhas com chaves nulas (Data Admissão/Jornada).
2.  **Normalização:** Separação de atributos descritivos (texto) em tabelas de dimensão, reduzindo a redundância na tabela fato.
3.  **Surrogate Keys:** Criação de IDs numéricos sequenciais (`id_servidor`, `id_cargo`, etc.) para substituir strings longas e otimizar joins no banco.
4.  **Enriquecimento:** Criação de colunas derivadas como `semestre`, `trimestre` e flags de atividade.

## 3. Destino dos Dados
Todas as tabelas foram persistidas no banco de dados **PostgreSQL** definido no arquivo `.env`. Os dados agora estão prontos para serem consumidos por ferramentas de BI (Power BI, Metabase) ou para as análises SQL comparativas.